api/desecapi/models.py

```diff
@@ -227,3 +230,68 @@ def save(self, *args, **kwargs):
 
     class Meta:
         ordering = ('created',)
+
+
+class RRset(models.Model):
+    # TODO Do these two fields really make sense? Meaning is limited when deleting + recreating an RRset
+    created = models.DateTimeField(auto_now_add=True)
+    updated = models.DateTimeField(null=True)
+    domain = models.ForeignKey(Domain, on_delete=models.CASCADE, related_name='rrsets')
+    subname = models.CharField(max_length=178, blank=True)
+    type = models.CharField(max_length=10)
+    records = models.CharField(max_length=64000, blank=True)
+    ttl = models.PositiveIntegerField(validators=[MinValueValidator(1)])
+    _dirty = False
+
+    class Meta:
+        unique_together = (("domain","subname","type"),)
+
+    def __setattr__(self, attrname, val):
+        setter_func = 'setter_' + attrname
+        if attrname in self.__dict__ and callable(getattr(self, setter_func, None)):
+            super().__setattr__(attrname, getattr(self, setter_func)(val))
+        else:
+            super().__setattr__(attrname, val)
+
+    def setter_records(self, val):
+        if val != self.records:
+            self._dirty = True
+
+        return val
+
+    def setter_ttl(self, val):
+        if val != self.ttl:
+            self._dirty = True
+
+        return val
+
+    def update_pdns(self):
+        from .serializers import RRsetSerializer
```

Comment:
I refactored the serialization logic for the use case of the API talking to pdns, see https://github.com/desec-io/desec-stack/commit/67f0bb19f508967f318d977ad72be2a8f0afff62. Everything is now done in `pdns.get_rrsets()` and `pdns.set_rrsets()` where `GenericRRsetSerializer` is used.

`GenericRRsetSerializer` is a `ModelSerializer`-based serializer. (It is extended by `RRsetSerializer` which is used by the frontend views.) The serializer transforms between `RRset` objects and a Python dictionary whose structure is driven by `ModelSerializer`.

`pdns.get_rrsets()` and `pdns.set_rrsets()` still have some list comprehension logic which is used to transform between the structure received from pdns and the Python dictionary produced by `GenericRRsetSerializer`.

At this point, we face two problems:

1. The list comprehension could be removed if the pdns JSON structure could be reproduced based on a `ModelSerializer` (such as `GenericRRsetSerializer`). However, it is not straightforward to write such a serializer, due to the (in our pdns usecase, constant but mandatory) extra fields in the pdns JSON structure (e.g. `'changetype': 'REPLACE'`). Note that this has nothing to do with whether `records` (list of strings) is deserialized into separate record objects or not.

   I am not sure if the extra work of making the serializer pdns-compatible is justified by the little gain of getting rid of the list comprehension. (I would take a different stand if this functionality was reused elsewhere in the code.)

2. Even with all transformations encapsulated in `pdns.get_rrsets()`, `pdns.set_rrsets()`, and `serializers.py`, a circular import issue occurs: `RRset.save()` calls `RRset.update_pdns()` which calls functions from `pdns.py`, so we need to import `pdns.py`. In `pdns.py`, we use `GenericRRsetSerializer`. Even if Problem 1 was solved, we would still need to call some serializer for `RRset` here. In that serializer's `Meta.model` attribute, the `RRset` model is referenced (as long as the serializer is a `ModelSerializer`), closing the loop.

   Summary: If within `RRset`, we call something that serializes an `RRset` object, and if the serializer used in that process is a `ModelSerializer`, then we have an import loop.

   Therefore, `pdns.get_rrsets()` and `pdns.set_rrsets()` have local imports at the moment, and I don't see a way of getting rid of them (except not using `ModelSerializer`). On the other hand, I don't see why a local import would be a problem. (The loop is not a logical inconsistency; it has merely to do with the fact that global import statements are not handled very smartly.)